charmcraft/commands/build.py

```diff
@@ -0,0 +1,178 @@
+    """Main entry point."""
+    buildpath = charmdir / BUILD_DIRNAME
+    logger.debug("Building charm in %r", str(buildpath))
+
+    # XXX Facundo 2020-05-18: for now we're re-building always from scratch, in the
+    # future we *may* reuse stuff already inside
+    if buildpath.exists():
+        shutil.rmtree(str(buildpath))
+    buildpath.mkdir()
+
+    # basic files
+    logger.debug("Linking in basic files and charm code")
+    link(charmdir / CHARM_METADATA, buildpath)
+
+    # the charm code
+    # XXX Facundo 2020-05-18: for now, one file, may we copy the entrypoint and its whole dir/tree
+    # (if not charmdir); also we need to understand if we want to ignore some files (e.g. .pyc)
+    linked_entrypoint = link(entrypoint, buildpath / 'src')
+
+    # dispatch mechanism
+    logger.debug("Creating the dispatch mechanism")
+    dispatch_content = DISPATCH.format(
+        entrypoint_relative_path=linked_entrypoint.relative_to(buildpath))
+    dispatch_path = buildpath / DISPATCH_FILENAME
+    with open(dispatch_path, "wt", encoding="utf8") as fh:
+        fh.write(dispatch_content)
+    os.chmod(dispatch_path, dispatch_path.stat().st_mode | stat.S_IXUSR)
+
+    # bunch of symlinks, to support old juju
+    hookpath = buildpath / 'hooks'
+    hookpath.mkdir()
+    for hookname in HOOK_NAMES:
+        (hookpath / hookname).symlink_to(dispatch_path)
+
+    # dependencies
+    logger.debug("Installing dependencies")
+    # XXX Facundo 2020-05-18: we may want to be flexible with how to include the
+    # dependencies, e.g. respecting current lib directory, or not having a requirements file
+    libpath = buildpath / 'lib'
+    reqspath = charmdir / 'requirements.txt'
+    cmd = [
+        'pip3', 'install',  # base command
+        '--system',  # indicates to use the system file structure
+        '--target={}'.format(libpath),  # put all the resulting files in that specific dir
+        '--requirement={}'.format(reqspath),  # the dependencies file
+    ]
+    retcode = polite_exec(cmd)
+    if retcode:
+        raise CommandError("problems installing the dependencies")
+
+    # z-z-zip it!
+    logger.debug("Creating the package itself")
+    zipname = charmdir.name + '.charm'
+    zipfh = zipfile.ZipFile(zipname, 'w', zipfile.ZIP_DEFLATED)
+    for dirpath, dirnames, filenames in os.walk(buildpath):
+        dirpath = pathlib.Path(dirpath)
+        for filename in filenames:
+            filepath = dirpath / filename
+            zipfh.write(str(filepath), str(filepath.relative_to(buildpath)))
+
+    zipfh.close()
+    logger.info("Done, charm left in %r", zipname)
+
+
+class BuildCommand(BaseCommand):
+    """Show the version."""
+    name = 'build'
+    help_msg = "build the charm"
+
+    def fill_parser(self, parser):
+        """Add own parameters to the general parser."""
+        parser.add_argument(
+            'charmdir', metavar='charm-dir',
+            help='The directory where the charm project is located')
```

Comment:
Couldn't this have a default of '.' ?
And entrypoint could have a default of 'src/charm.py' ?

Reply:
Yes, added that in later commit.